File: built-in plugins dsp/mb-bass-acid.cpp
/**
 * MB Acid Bass
 * Category : instrument
 * Type     : bass
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : TB-303 style acid bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BASS_ACID_H
#define MB_BASS_ACID_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBassAcid : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bass-acid";
    static constexpr const char* PLUGIN_NAME    = "MB Acid Bass";
    static constexpr const char* PLUGIN_TYPE    = "bass";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float cutoff = 2000f;  // range [100, 10000]
    float resonance = 0.7f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbBassAcid() = default;
    ~MbBassAcid() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.cutoff = std::clamp(params.cutoff, 100f, 10000f);
        params.resonance = std::clamp(params.resonance, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Acid Bass
        return input;
    }
};

#endif // MB_BASS_ACID_H

File: built-in plugins dsp/mb-bass-acid.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbBassAcidPlugin: PluginDefinition = {
    id: 'mb-bass-acid', slug: 'mb-bass-acid', name: 'MB Acid Bass', category: 'instrument', type: 'bass', version: '1.0.0',
    description: 'TB-303 style acid bass', author: 'Max Booster',
    oscillators: [{ type: 'sawtooth', detune: 0, gain: 0.8 }, { type: 'square', detune: 0, gain: 0.2 }],
    envelope: { attack: 0.001, decay: 0.3, sustain: 0.3, release: 0.1 },
    parameters: [
      { id: 'cutoff', name: 'Cutoff', type: 'float', defaultValue: 2000, minValue: 100, maxValue: 10000, automatable: true },
      { id: 'resonance', name: 'Resonance', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { cutoff: 2000, resonance: 0.7, volume: 0.8 },
  };

export default MbBassAcidPlugin;
